fvtest/compilertriltest/LinkageTest.cpp

```diff
@@ -151,3 +150,291 @@ TYPED_TEST(LinkageTest, SystemLinkageParameterPassingFourArg) {
+    // range for some architectures.
+#ifdef TR_TARGET_X86
+    Tril::DefaultCompiler compiler{trees};
+    ASSERT_EQ(0, compiler.compile()) << "Compilation failed unexpectedly\n" << "Input trees: " << inputTrees;
+
+
+    auto entry_point = compiler.getEntryPoint<TypeParam (*)(TypeParam,TypeParam,TypeParam,TypeParam,TypeParam)>();
+
+    EXPECT_EQ(static_cast<TypeParam>(2053),     stackUser<TypeParam>(1,1,1,1,static_cast<TypeParam>(1024)))     << "Input Trees: " << inputTrees;
+    EXPECT_EQ(static_cast<TypeParam>(3),        stackUser<TypeParam>(1,1,1,1,static_cast<TypeParam>(-1)))       << "Input Trees: " << inputTrees;
+    EXPECT_EQ(static_cast<TypeParam>(0x1e1e23), stackUser<TypeParam>(1,1,1,1,static_cast<TypeParam>(0xf0f0f)))  << "Input Trees: " << inputTrees;
+
+    EXPECT_EQ(static_cast<TypeParam>(2053),     entry_point(1,1,1,1,static_cast<TypeParam>(1024)))     << "Input Trees: " << inputTrees;
+    EXPECT_EQ(static_cast<TypeParam>(3),        entry_point(1,1,1,1,static_cast<TypeParam>(-1)))       << "Input Trees: " << inputTrees;
+    EXPECT_EQ(static_cast<TypeParam>(0x1e1e23), entry_point(1,1,1,1,static_cast<TypeParam>(0xf0f0f)))  << "Input Trees: " << inputTrees;
+#endif
+}
+
+class LinkageWithMixedTypesTest : public TRTest::JitTest {};
+
+int fourthArgFromMixedTypes(double a, int b, double c, int d) { return d; }
+
+typedef int (*FourMixedArgumentFunction)(double,int,double,int);
+
+TEST_F(LinkageWithMixedTypesTest, SystemLinkageParameterPassingFourArgWithMixedTypes) {
+    char inputTrees[400] = {0};
+    const auto format_string = "(method return=Int32 args=[Double,Int32,Double,Int32]"
+                               "  (block (ireturn (icall address=0x%jX args=[Double,Int32,Double,Int32] linkage=system"
+                                 " (dload parm=0)"
+                                 " (iload parm=1)"
+                                 " (dload parm=2)"
+                                 " (iload parm=3)"
+                                 ") )  ))";
+    std::snprintf(inputTrees, 400, format_string,
+        reinterpret_cast<uintmax_t>(static_cast<FourMixedArgumentFunction>(fourthArgFromMixedTypes)));
+
+    auto trees = parseString(inputTrees);
+    ASSERT_NOTNULL(trees) << "Trees failed to parse\n" << inputTrees;
+
+    // Execution of this test is disabled on non-X86 platforms, as we
+    // do not have trampoline support, and so this call may be out of
+    // range for some architectures.
+#ifdef TR_TARGET_X86
+    Tril::DefaultCompiler compiler{trees};
+    ASSERT_EQ(0, compiler.compile()) << "Compilation failed unexpectedly\n" << "Input trees: " << inputTrees;
+
+    auto entry_point = compiler.getEntryPoint<FourMixedArgumentFunction>();
+
+    EXPECT_EQ(1024,    entry_point(0.0,0,0.0,1024))     << "Input Trees: " << inputTrees;
+    EXPECT_EQ(-1,      entry_point(0.0,0,0.0,-1))       << "Input Trees: " << inputTrees;
+    EXPECT_EQ(0xf0f0f, entry_point(0.0,0,0.0,0xf0f0f))  << "Input Trees: " << inputTrees;
```

Comment:
Would it be possible to turn this into a value-parameterize test?

Reply:
@Leonardo2718 I use different values for both tests in `LinkageWithMixedTypesTest` since one parameter list ends with integer while the other ends with double. There could be two value-parameterized tests or I can pass a vector of values (but with different types and it could make the tests more compilcated) but I see no reason for it: the tests will be more difficult for reading and understanding what is being testing there.

Reply:
I see, you're right.

As an alternative then to increase input coverage, how about something like:

```c++
    auto inputs = TRTest::const_values<int32_t>();
    for (auto i = inputs.cbegin(); i != inputs.cend(); ++i) {
        SCOPED_TRACE(*i);
        EXPECT_EQ(*i, entry_point(0.0,0,0.0, *i))  << "Input Trees: " << inputTrees;
    }
```

Reply:
@Leonardo2718 I've defined two additional `const_values` template implementations: 

```cpp
template <>
inline std::vector<float> const_values<float>()
```

and

```cpp
template <>
inline std::vector<double> const_values<double>()
```

and changed the three test 'expections' in every test to a check against each const_value except the `SystemLinkageParameterPassingFiveArgToStackUser` test (because the expected value is not matched to the last argument).

```cpp
    auto inputs = TRTest::const_values<TypeParam>();
    for (auto i = inputs.cbegin(); i != inputs.cend(); ++i) {
        SCOPED_TRACE(*i);
        EXPECT_EQ(static_cast<TypeParam>(*i), entry_point(0,0,0,0, *i))  << "Input Trees: " << inputTrees;
    }
```